### ISSUE TITLE
Extension is running ansible-lint on all yaml files, causing high CPU usage

### ISSUE THREAD

**Issue by aliusmiles** [ansible-community/vscode-ansible#58]
This extension is spawning tons of Python processes by running ansible-lint on all yaml files in all projects, including kubernetes manifests, etc. Results in huge CPU usage and high load. Found by running `pgrep -fl Python` after opening some k8s project.

Fixed for now by `"ansible.validate.enable": false` in settings.json.

Could it be configured to run lint only on 'ansible' files and not on all yaml?

I've noticed that extension is not "defining" 'ansbile' language in package.json, as in now deprecated [vscoss.vscode-ansible extension](https://github.com/VSChina/vscode-ansible/blob/master/package.json) but rather extending 'YAML' language, might that be root of this issue? Or it's just some misconfiguration on my system?

I'm happy to provide any additional info if needed.

VSCode: 1.55.0
zbr.vscode-ansible: v0.3.0
OS: MacOS 11.2.3

**Comment by ssbarnea** [comment]
You are likely facing a problem from another JSON schema being used. As you correctly observed we only hint YAML extension about which YAML files are ansible related, but the legwork is done bit YAML language-server.

This means that the files causing heavy load may be files unrelated to ansible. This happened to me in the past and I sorted it by adding `"yaml.schemaStore.enable": false,` to settings.json file. As you can see, this will disable looking for other schemas not explicitly configured by user or declared by extensions like ours.

**Comment by aliusmiles** [comment]
I've tried `"yaml.schemaStore.enable": false,` - doesn't help, still running ansible-lint on all yaml files. 

Spawned processes looks like following:

```sh
pgrep -fl Python

82400 /usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/Resources/Python.app/Contents/MacOS/Python /usr/local/bin/ansible-lint --nocolor -p /Users/taras/Projects/test/base/configmap.yml
```

Also, default config is to run validation on save, but all those processes starts immediately after I open project/workspace with yaml files and for all yaml files in project, not only opened.

Checked Ansible extension output, got only 'Ansible Language extension is now active'. Any way to enable debug logs perhaps?

**Comment by ssbarnea** [comment]
Not really. Out of curiosity can you time the operation? It should not take more than 2-3s to lint one file.

**Comment by aliusmiles** [comment]
@ssbarnea it does take 2-3s on single file, but when I'm opening workspace with multiple repos and about 500 large yaml files and 500 corresponding asible-lint processes are spawned - it does finish all lint processes in about 10 sec, but load average got to 400:) and it takes couple min for everything to get to normal (fan noise, temp, unresponsive vscode)

It's quite easy to reproduce this issue, at least on my setup:

settings.json: `"ansible.validate.enable": false,`

```sh
for i in {1..100}; do echo -e "---\ntest: test" > ${i}.yaml; done
code .
pgrep -fl Python
```

I did test it with all extensions disabled, except zbr.vscode-ansible, redhat.vscode-yaml, ms-python.python and ms-toolsai.jupyter ( enabled by python I think) - ansible-lint is still executed on all yaml files and not only ansible ones, but interestingly it runs only on opened files one at a time and not on all yamls in workspace.

So looks like 1) enabling ansible.validate makes ansible-lint to run agains any yaml files and not only "ansible" ones, and 2) some other extension on my setup clearly makes it run against all yaml files in workspace.

Could you maybe try to reproduce? At this point I'm not sure if issue 1) is indeed on extension's side or just something wrong with my setup.

**Comment by aliusmiles** [comment]
Here's list of my extensions:
```sh
adamhartford.vscode-base64
ahmadalli.vscode-nginx-conf
alefragnani.Bookmarks
alefragnani.project-manager
amazonwebservices.aws-toolkit-vscode
bajdzis.vscode-database
bungcip.better-toml
chunsen.bracket-select
DavidAnson.vscode-markdownlint
eamodio.gitlens
felipe.nasc-touchbar
GitHub.github-vscode-theme
golang.go
GrapeCity.gc-excelviewer
Gruntfuggly.global-config
Gruntfuggly.todo-tree
hangxingliu.vscode-nginx-conf-hint
hashicorp.terraform
hediet.vscode-drawio
HookyQR.beautify
IBM.output-colorizer
ivory-lab.jenkinsfile-support
janjoerke.jenkins-pipeline-linter-connector
JuanBlanco.solidity
mads-hartmann.bash-ide-vscode
mechatroner.rainbow-csv
mindaro-dev.file-downloader
ms-azuretools.vscode-docker
ms-kubernetes-tools.vscode-kubernetes-tools
ms-python.python
ms-python.vscode-pylance
ms-toolsai.jupyter
ms-vscode-remote.remote-containers
ms-vscode-remote.remote-ssh
ms-vscode-remote.remote-ssh-edit
ms-vscode-remote.remote-ssh-explorer
ms-vscode-remote.remote-wsl
ms-vscode-remote.vscode-remote-extensionpack
ms-vscode.hexeditor
PKief.material-icon-theme
raynigon.nginx-formatter
redhat.vscode-yaml
ryu1kn.partial-diff
shd101wyy.markdown-preview-enhanced
sidneys1.gitconfig
stubailo.ignore-gitignore
ThreadHeap.serverless-ide-vscode
vscodevim.vim
wholroyd.jinja
william-voyek.vscode-nginx
wingrunr21.vscode-ruby
wwm.better-align
yzhang.markdown-all-in-one
zbr.vscode-ansible
```

**Comment by ssbarnea** [comment]
My impression was that only the opened files were validated. The linter itself can process files recursively and the speed gains are significant.

In fact the ~2s delays comes from ansible own bootstrapping time, and the linter does that for each file (still running in parallel).

While for standalone cli this was a known issue, it seems to become more problematic when it comes to using it inside a live editor.

**Comment by aliusmiles** [comment]
Well, with only ansible and yaml extensions enabled it's indeed validating only opened files for me, but with all extensions it's validating all the yaml files from workspace as soon as I open any yaml file 🤷‍♂️ Must be something in my setup though no idea how to debug this.

And it seems to me that it should validate only ansible files, like tasks/main.yaml and not all yaml files like test.yaml, unless that is caused by something on my side as well.

So far I'm quite happy to just disable validation since molecule takes care of it for me, but if my assumptions are correct then ansible extension does not work as intended with validation. Unfortunately I don't know Typescript enough to investigate further, but I'm happy to provide logs or any other info if needed.

**Comment by ssbarnea** [comment]
That is indeed a very valuable insight. Clearly one of the listed extensions does produce this undersired behavior, likely one that opens yaml files somehow. By looking at the list I only see `wholroyd.jinja` but also a bunch of other generic ones. Please try to identify the culprit by disabling them, you can use the [bisecting feature](https://code.visualstudio.com/blogs/2021/02/16/extension-bisect) to speedup your search.

**Comment by aliusmiles** [comment]
Thanks for suggestion on bisect, very useful! (for reference, had to disable ansible/yaml extensions before starting bisect and then enable in order to exclude them from bisect)

And extension that caused parallel runs of ansible-lint is [threadheap.serverless-ide-vscode](https://github.com/threadheap/serverless-ide-vscode)

I don't use it mach now if at all so I've simply uninstalled it. Any advise on troubleshooting steps, just in case?

Guess this issue can be closed now, although I'm not quite sure if running ansible-lint on any opened yaml file and not only ansible ones is a desired behavior or a bug? Should I perhaps create another issue for that?

**Comment by ssbarnea** [comment]
@aliusmiles Be sure that you create a bug with them and link with it from here. In fact redhat-yaml extension has the ability to report incompatible extensions and ask user to uninstall them.

@JPinkney should we add one more extension to the banlist?

**Comment by JPinkney** [comment]
It's hard to say. I went through their repo and it looks like they're running their own custom language server against yaml files and not just re-using the yaml-language-server. The other extensions in the banlist were just re-using the yaml language server IIRC. I'd probably just open an issue on the serverless vscode extension's github repo with your investigation and see what comes out of it

**Comment by ssbarnea** [comment]
Closing because that is caused by external plugin.

**Comment by andreabolandrina** [comment]
Hi,

I was having multiple ansible-lint processes too (not on all yaml files though, just the one I touch).

This caused high CPU and, above all, high I/O, making the system (a mere RPI-3) non responsive.

For now I've disable the linting by accessing the extension settings and unticking the "Ansible › Validation › Lint: Enabled
" setting.

Hope it helps.

Andrea

**Comment by ssbarnea** [comment]
@andreabolandrina I am not sure if using a RPI was ever considered as suitable for ansible-lint. Running on all files is by design as change on one file can produce failures on other ones.

**Comment by andreabolandrina** [comment]
"I am not sure if using a RPI was ever considered as suitable for ansible-lint" I thought so too, and that is why I've specified what I was using.

If somebody else is having the same issue though, my solution above might prove useful as a workaround (until they/we upgrade to something more suitable).

**Comment by ssbarnea** [comment]
I plan to add some features that would reduce the load on re-linting actions, like a configurable delay. That combined with upcoming improvements on performance, will make it suitable for RPI too.

**Comment by ad0nis** [comment]
I am also experiencing this issue on a minimum-spec box in Digital Ocean. Large ansible project with many files, and it appears to be spinning up the linter for every one of them on every change. Completely locks the box with 100% cpu and 100% memory usage. Half the time the tests crash and I lose my connections to the system and have to start over, wasting time.

**Comment by ssbarnea** [comment]
@ad0nis "minimum-spec box" should hint you about the fix. Performance was never no1 priority for the linter. I know for sure that we will improve the performance considerably later this year but for the moment you will need to scale up the hardware a little bit.

**Comment by ad0nis** [comment]
@ssbarnea : You're absolutely right. I had seen users commenting on having issues with a raspberry pi in this thread, and figured this might be another useful data-point, as it's another entry-level development solution I assume might be common.

That said, I was able to work around this eventually. -- I had found multiple guides referencing removing the directory the teamserver is installed in on the server ( https://earlruby.org/2021/06/fixing-vscode-when-it-keeps-dropping-ssh-connections/ ) and went ahead and did that, and allowed it to do a fresh re-install. I also added some swap space to the box and that's removed the memory bottleneck, which seems to have made the system overall more reliable when using vscode along with the ansible linter. It still spikes the CPU, but it feels like it's dropping less connections.

Will look into the py311 changes, but had heard rumblings about 3.10 breaking various crypto tools, and haven't touched it yet a as result. Those speed increases do seem worth it though...
